[FLINK-3937] programmatic resuming of clusters

mxm opened the pull request:
These changes are based on #1978 and #2034. More specifically, they port resuming of running Yarn clusters from #2034 to the refactoring of #1978. The abstraction in place enables us to plug in other cluster frameworks in the future. 

- integrates with and extends the refactoring of FLINK-3667
- enables to resume from Yarn properties or Yarn application id
- introduces additional StandaloneClusterDescriptor
- introduces DefaultCLI to get rid of standalone mode switches in CliFrontend
- various fixes and improvements

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java:
```diff
@@ -980,110 +845,41 @@ protected ActorGateway getJobManagerGateway(CommandLineOptions options) throws E
-			}
+		// Get the custom command-line (e.g. Standalone/Yarn/Mesos)
+		CustomCommandLine<?> activeCommandLine =
+			CliFrontendParser.getActiveCustomCommandLine(options.getJobManagerAddress());
```
Why is this method located in the Parser?
I thought the parser is only responsible for parsing the command line arguments?

Maybe it makes more sense to rename this 
```
loadCustomCommandLine("org.apache.flink.client.cli.DefaultCLI");
		loadCustomCommandLine("org.apache.flink.yarn.cli.FlinkYarnSessionCli", "y", "yarn");
```
to parser (loadCustomParser(), and DefaultCLIParser, YarnCliParser?)

mxm replied:
Well spotted. I think I'll move this logic to the implementation of the CustomCommandLine. 

I don't quite understand your renaming suggestion, are you suggesting to break up the CustomCommandLine into CustomParser and CustomCLI?

rmetzger replied:
Okay, moving the logic is addressing my concerns.

I checked the classes again and my renaming suggestions don't make much sense ;)

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java:
```diff
@@ -980,110 +845,41 @@ protected ActorGateway getJobManagerGateway(CommandLineOptions options) throws E
-				writeJobManagerAddressToConfig(jobManagerAddress);
+		if (client != null) {
+			logAndSysout("Cluster retrieved");
+		} else {
```
I wonder if it would be a bit more obvious to have some methods on the `ClusterClient` to check if they can retrieve a cluster or if they need to create one.
Right now its a bit implicit through the `null`.

mxm replied:
Yes, I have thought about letting the command-line decide whether a cluster can be created or retrieved (it would return an Enum based on the options supplied). I decided I would address this in a follow-up.

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java:
```diff
@@ -1275,33 +1071,16 @@ else if (new File(CONFIG_DIRECTORY_FALLBACK_2).exists()) {
+	 * @param address Address to write to the configuration
+	 * @param config The config to write to
+	 */
+	public static void writeJobManagerAddressToConfig(Configuration config, InetSocketAddress address) {
```
How about renaming this to "setJobManagerAddressInConfig()" .. write implies that something is written to a file or something.

mxm replied:
+1

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/cli/CustomCommandLine.java:
```diff
@@ -0,0 +1,65 @@
+	 * Retrieves a client for a running cluster
+	 * @param commandLine The command-line parameters from the CliFrontend
+	 * @param config The Flink config
+	 * @return Client if a cluster could be retrieve, null otherwise
```
typo: retrieved

mxm replied:
Thanks.

rmetzger commented:
I quickly tried your changes.
Some notes
(1)  When listing the yarn-session.sh cli options, I could not see the newly added argument
```
./bin/yarn-session.sh 
2016-06-10 09:28:37,186 ERROR org.apache.flink.yarn.cli.FlinkYarnSessionCli                 - Missing required argument n
Usage:
   Required
     -n,--container <arg>   Number of YARN container to allocate (=Number of Task Managers)
   Optional
     -D <arg>                        Dynamic properties
     -d,--detached                   Start detached
     -jm,--jobManagerMemory <arg>    Memory for JobManager Container [in MB]
     -nm,--name <arg>                Set a custom name for the application on YARN
     -q,--query                      Display available YARN resources (memory, cores)
     -qu,--queue <arg>               Specify YARN queue.
     -s,--slots <arg>                Number of slots per TaskManager
     -st,--streaming                 Start Flink in streaming mode
     -tm,--taskManagerMemory <arg>   Memory per TaskManager Container [in MB]
Error while starting the YARN Client. Please check log output!
```

(2) Starting a YARN session works. Also, connecting with a second YARN session works. However, if I'm stopping the session from the second session client, the first one does not notice this.
It would be good if all connected clients receive the messages, instead of the last one connected.

rmetzger commented:
I'll probably continue reviewing your work next Monday.

rmetzger commented on flink-clients/src/main/java/org/apache/flink/client/program/ClusterClient.java:
```diff
@@ -181,11 +198,30 @@ public boolean getPrintStatusDuringExecution() {
+		try {
+		String hostName = flinkConfig.getString(ConfigConstants.JOB_MANAGER_IPC_ADDRESS_KEY, null);
+		int port = flinkConfig.getInteger(ConfigConstants.JOB_MANAGER_IPC_PORT_KEY, -1);
+		return new InetSocketAddress(hostName, port);
```
Missing indentation

mxm replied:
Fixed.

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClient.java:
```diff
@@ -211,18 +185,45 @@ public void run() {
 
 		isConnected = true;
+
+		logAndSysout("Waiting until all TaskManagers have connected");
```
Since the `logAndSysout` method is defined in the `ClusterClient`, all log messages appear from being from that class, even though they are coming from the `YarnClusterClient`. Maybe it makes sense to override the method or add an argument for the logger.

mxm replied:
Thank you. I fixed that by using a non-static variable for the logger and dynamically retrieving the class name, i.e.
```java
	private final Logger LOG = LoggerFactory.getLogger(getClass());
```

rmetzger commented on docs/apis/cli.md:
```diff
@@ -105,6 +105,10 @@ The command line can be used to
 
+-   List running Flink jobs inside Flink YARN session:
+
+        ./bin/flink list -m yarn-cluster -yid <yarnApplicationID> -r
```
Why is the client trying to connect even though the application has been finished already?

```
[cloudera@quickstart build-target]$  ./bin/flink list -m yarn-cluster -yid application_1447844011707_0036 -r
2016-06-13 06:51:34,581 INFO  org.apache.hadoop.yarn.client.RMProxy                         - Connecting to ResourceManager at quickstart.cloudera/10.0.2.15:8032
2016-06-13 06:51:35,017 ERROR org.apache.flink.yarn.YarnClusterDescriptor                   - The application application_1447844011707_0036 doesn't run anymore. It has previously completed with final status: SUCCEEDED

------------------------------------------------------------
 The program finished with the following exception:

org.apache.flink.runtime.leaderretrieval.LeaderRetrievalException: Could not retrieve the leader gateway
	at org.apache.flink.runtime.util.LeaderRetrievalUtils.retrieveLeaderGateway(LeaderRetrievalUtils.java:127)
	at org.apache.flink.client.CliFrontend.getJobManagerGateway(CliFrontend.java:844)
	at org.apache.flink.client.CliFrontend.list(CliFrontend.java:378)
	at org.apache.flink.client.CliFrontend.parseParameters(CliFrontend.java:986)
	at org.apache.flink.client.CliFrontend.main(CliFrontend.java:1034)
Caused by: java.util.concurrent.TimeoutException: Futures timed out after [10000 milliseconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
	at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:107)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.Await$.result(package.scala:107)
	at scala.concurrent.Await.result(package.scala)
	at org.apache.flink.runtime.util.LeaderRetrievalUtils.retrieveLeaderGateway(LeaderRetrievalUtils.java:125)
	... 4 more
```

rmetzger replied:
This is a similar thing:

```
[cloudera@quickstart build-target]$  ./bin/flink list -m yarn-cluster -yid obiouslyWrong -r

------------------------------------------------------------
 The program finished with the following exception:

org.apache.flink.runtime.leaderretrieval.LeaderRetrievalException: Could not retrieve the leader gateway
	at org.apache.flink.runtime.util.LeaderRetrievalUtils.retrieveLeaderGateway(LeaderRetrievalUtils.java:127)
	at org.apache.flink.client.CliFrontend.getJobManagerGateway(CliFrontend.java:844)
	at org.apache.flink.client.CliFrontend.list(CliFrontend.java:378)
	at org.apache.flink.client.CliFrontend.parseParameters(CliFrontend.java:986)
	at org.apache.flink.client.CliFrontend.main(CliFrontend.java:1034)
Caused by: java.util.concurrent.TimeoutException: Futures timed out after [10000 milliseconds]
	at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:219)
	at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:223)
	at scala.concurrent.Await$$anonfun$result$1.apply(package.scala:107)
	at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
	at scala.concurrent.Await$.result(package.scala:107)
	at scala.concurrent.Await.result(package.scala)
	at org.apache.flink.runtime.util.LeaderRetrievalUtils.retrieveLeaderGateway(LeaderRetrievalUtils.java:125)
	... 4 more
[cloudera@quickstart build-target]$ 
```

rmetzger replied:
Are there any tests for this functionality?

mxm replied:
Thank you, I'll look into this.

mxm replied:
This is handled correctly now.

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -415,62 +513,83 @@ public int run(String[] args) {
 			}
 			System.out.println(description);
 			return 0;
+		} else if (cmd.hasOption(APPLICATION_ID.getOpt())) { // TODO RM
```
What is this TODO about?

mxm replied:
Sorry, added this TODO while I fixing some of your first comments.

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java:
```diff
@@ -108,24 +131,83 @@ public FlinkYarnSessionCli(String shortPrefix, String longPrefix, boolean accept
+	public YarnClusterDescriptor createDescriptor(String defaultApplicationName, CommandLine cmd) {
+
+
+		YarnClusterDescriptor yarnClusterDescriptor = new YarnClusterDescriptor();
```
A bit too many blank lines ;)

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterClient.java:
```diff
@@ -211,18 +185,45 @@ public void run() {
+
+		logAndSysout("Waiting until all TaskManagers have connected");
+
+		while(true) {
```
It seems that this loop is now executed also for the YARN session. In the past, this was used only for "per job" yarn clusters, because there we needed to have all slots available to submit the job.

Now, this loop is "blocking" the YARN session CLI if not all TaskManager's are connecting. Its not a big deal, but a change in the semantics.

mxm replied:
Yes, I wanted to bring up the Yarn session only once the cluster is ready. It is a semantic change but IMHO transparent to the user. There is not disadvantage from waiting until the cluster is ready. Ultimately, it would be nice to get rid of all waiting but we're not quite there yet.

rmetzger replied:
I agree that the functionality is the same.

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -300,18 +309,82 @@ public static boolean allocateResource(int[] nodeManagers, int toAllocate) {
+					return retrieve(report.getApplicationId().toString());
+				}
+			}
+
```
I wonder if it makes sense to log here that we were unable to find the application id?

mxm replied:
Good idea.

rmetzger commented on flink-yarn/src/main/java/org/apache/flink/yarn/AbstractYarnClusterDescriptor.java:
```diff
@@ -300,18 +309,82 @@ public static boolean allocateResource(int[] nodeManagers, int toAllocate) {
+			}
+
+		}
+		return null;
```
Does it make sense to log that the config hasn't been set?

mxm replied:
Same as above. Fixed.

rmetzger commented on docs/apis/cli.md:
```diff
@@ -252,6 +256,12 @@ Action "list" lists running and scheduled programs.
+                                   connect to. Must not be set if JobManager HA
+                                   is used. In this case, JobManager RPC
+                                   location is automatically retrieved from
+                                   Zookeeper.
```
I tried this, but I wonder why its not logging that the job has been cancelled. 
It logs that "All TaskManagers are connected", I don't think this message is relevant when cancelling a job.

```
[cloudera@quickstart build-target]$  ./bin/flink cancel -m yarn-cluster -yid application_1447844011707_0038 b9b8f76616073d09c596545a3cda978f
2016-06-13 07:22:21,355 INFO  org.apache.hadoop.yarn.client.RMProxy                         - Connecting to ResourceManager at quickstart.cloudera/10.0.2.15:8032
2016-06-13 07:22:22,277 INFO  org.apache.flink.yarn.YarnClusterClient                       - Start application client.
2016-06-13 07:22:22,288 INFO  org.apache.flink.yarn.ApplicationClient                       - Notification about new leader address akka.tcp://flink@10.0.2.15:51747/user/jobmanager with session ID null.
2016-06-13 07:22:22,290 INFO  org.apache.flink.yarn.ApplicationClient                       - Received address of new leader akka.tcp://flink@10.0.2.15:51747/user/jobmanager with session ID null.
2016-06-13 07:22:22,290 INFO  org.apache.flink.yarn.ApplicationClient                       - Disconnect from JobManager null.
Waiting until all TaskManagers have connected
2016-06-13 07:22:22,297 INFO  org.apache.flink.yarn.ApplicationClient                       - Trying to register at JobManager akka.tcp://flink@10.0.2.15:51747/user/jobmanager.
No status updates from the YARN cluster received so far. Waiting ...
2016-06-13 07:22:22,542 INFO  org.apache.flink.yarn.ApplicationClient                       - Successfully registered at the ResourceManager using JobManager Actor[akka.tcp://flink@10.0.2.15:51747/user/jobmanager#1733798764]
All TaskManagers are connected
2016-06-13 07:22:22,945 INFO  org.apache.flink.yarn.YarnClusterClient                       - Shutting down YarnClusterClient from the client shutdown hook
2016-06-13 07:22:22,945 INFO  org.apache.flink.yarn.YarnClusterClient                       - Disconnecting YarnClusterClient from ApplicationMaster
2016-06-13 07:22:22,947 INFO  org.apache.flink.yarn.ApplicationClient                       - Stopped Application client.
2016-06-13 07:22:22,947 INFO  org.apache.flink.yarn.ApplicationClient                       - Disconnect from JobManager Actor[akka.tcp://flink@10.0.2.15:51747/user/jobmanager#1733798764].
2016-06-13 07:22:23,056 INFO  org.apache.flink.yarn.YarnClusterClient                       - Application application_1447844011707_0038 finished with state RUNNING and final state UNDEFINED at 0
[cloudera@quickstart build-target]$ 
```

mxm replied:
Same as for `list` and `info`. I agree this verbosity is not very nice. Looking into how I can make this look nicer.

mxm replied:
The code no longer waits for TaskManagers when the cluster is resumed.

rmetzger commented:
While trying out your code, I noticed that the `.yarn-properties` file is not properly deleted, even though my previous yarn session was shutting down correctly:

```
2016-06-13 07:16:51,885 INFO  org.apache.flink.client.program.ClusterClient                 - TaskManager status (0/1)
TaskManager status (0/1)
2016-06-13 07:16:52,386 INFO  org.apache.flink.client.program.ClusterClient                 - TaskManager status (0/1)
TaskManager status (0/1)
2016-06-13 07:16:52,887 INFO  org.apache.flink.client.program.ClusterClient                 - All TaskManagers are connected
All TaskManagers are connected
2016-06-13 07:16:52,887 INFO  org.apache.flink.client.program.ClusterClient                 - Looking up JobManager
2016-06-13 07:16:52,912 INFO  org.apache.flink.client.program.ClusterClient                 - Looking up JobManager
Flink JobManager is now running on 10.0.2.15:51747
JobManager Web Interface: http://quickstart.cloudera:8088/proxy/application_1447844011707_0038/
Number of connected TaskManagers changed to 1. Slots available: 1
^[[A^C2016-06-13 07:25:35,370 INFO  org.apache.flink.yarn.YarnClusterClient                       - Shutting down YarnClusterClient from the client shutdown hook
2016-06-13 07:25:35,372 INFO  org.apache.flink.yarn.YarnClusterClient                       - Sending shutdown request to the Application Master
2016-06-13 07:25:35,373 INFO  org.apache.flink.yarn.ApplicationClient                       - Sending StopCluster request to JobManager.
2016-06-13 07:25:35,429 INFO  org.apache.flink.yarn.ApplicationClient                       - Stopped Application client.
2016-06-13 07:25:35,431 INFO  org.apache.flink.yarn.ApplicationClient                       - Disconnect from JobManager Actor[akka.tcp://flink@10.0.2.15:51747/user/jobmanager#1733798764].
2016-06-13 07:25:35,469 INFO  akka.remote.RemoteActorRefProvider$RemotingTerminator         - Shutting down remote daemon.
2016-06-13 07:25:35,469 INFO  akka.remote.RemoteActorRefProvider$RemotingTerminator         - Remote daemon shut down; proceeding with flushing remote transports.
2016-06-13 07:25:35,622 INFO  akka.remote.RemoteActorRefProvider$RemotingTerminator         - Remoting shut down.
2016-06-13 07:25:35,701 INFO  org.apache.flink.yarn.YarnClusterClient                       - Deleting files in hdfs://quickstart.cloudera:8020/user/cloudera/.flink/application_1447844011707_0038
2016-06-13 07:25:35,706 INFO  org.apache.flink.yarn.YarnClusterClient                       - Application application_1447844011707_0038 finished with state FINISHED and final state SUCCEEDED at 1465827935399
2016-06-13 07:25:36,567 INFO  org.apache.flink.yarn.YarnClusterClient                       - YARN Client is shutting down
(reverse-i-search)`yarn': ./bin/^Crn-session.sh -n 1
[cloudera@quickstart build-target]$ ./bin/flink run -y yarn-cluster -yd -yn 1 ./examples/batch/WordCount.jar 
JAR file does not exist: -y

Use the help option (-h or --help) to get help on the command.
[cloudera@quickstart build-target]$ ./bin/flink run -m yarn-cluster -yd -yn 1 ./examples/batch/WordCount.jar 
2016-06-13 07:26:11,057 INFO  org.apache.flink.yarn.cli.FlinkYarnSessionCli                 - Found YARN properties file /tmp/.yarn-properties-cloudera
2016-06-13 07:26:11,057 INFO  org.apache.flink.yarn.cli.FlinkYarnSessionCli                 - Found YARN properties file /tmp/.yarn-properties-cloudera
Found YARN properties file /tmp/.yarn-properties-cloudera
2016-06-13 07:26:11,082 INFO  org.apache.flink.yarn.cli.FlinkYarnSessionCli                 - Using JobManager address from YARN properties quickstart.cloudera/10.0.2.15:51747
2016-06-13 07:26:11,082 INFO  org.apache.flink.yarn.cli.FlinkYarnSessionCli                 - Using JobManager address from YARN properties quickstart.cloudera/10.0.2.15:51747
Using JobManager address from YARN properties quickstart.cloudera/10.0.2.15:51747
2016-06-13 07:26:11,270 INFO  org.apache.hadoop.yarn.client.RMProxy                         - Connecting to ResourceManager at quickstart.cloudera/10.0.2.15:8032
2016-06-13 07:26:11,795 INFO  org.apache.flink.yarn.YarnClusterDescriptor                   - Found application 'application_1447844011707_0038' with JobManager host name 'quickstart.cloudera' and port '51747' from Yarn properties file.
2016-06-13 07:26:11,842 INFO  org.apache.hadoop.yarn.client.RMProxy                         - Connecting to ResourceManager at quickstart.cloudera/10.0.2.15:8032
2016-06-13 07:26:11,878 ERROR org.apache.flink.yarn.YarnClusterDescriptor                   - The application application_1447844011707_0038 doesn't run anymore. It has previously completed with final status: SUCCEEDED

------------------------------------------------------------
 The program finished with the following exception:

java.lang.RuntimeException: The Yarn application application_1447844011707_0038 doesn't run anymore.
	at org.apache.flink.yarn.AbstractYarnClusterDescriptor.retrieve(AbstractYarnClusterDescriptor.java:377)
	at org.apache.flink.yarn.AbstractYarnClusterDescriptor.retrieveFromConfig(AbstractYarnClusterDescriptor.java:351)
	at org.apache.flink.yarn.cli.FlinkYarnSessionCli.retrieveCluster(FlinkYarnSessionCli.java:464)
	at org.apache.flink.yarn.cli.FlinkYarnSessionCli.retrieveCluster(FlinkYarnSessionCli.java:62)
	at org.apache.flink.client.CliFrontend.getClient(CliFrontend.java:862)
	at org.apache.flink.client.CliFrontend.run(CliFrontend.java:228)
	at org.apache.flink.client.CliFrontend.parseParameters(CliFrontend.java:983)
	at org.apache.flink.client.CliFrontend.main(CliFrontend.java:1034)
[cloudera@quickstart build-target]$ 
```

rmetzger commented:
Detached per job submission is not working (`./bin/flink run -m yarn-cluster -yd -yn 1 ./examples/batch/WordCount.jar `)

```
2016-06-13 07:28:22,423 INFO  org.apache.flink.yarn.ApplicationClient                       - Trying to register at JobManager akka.tcp://flink@10.0.2.15:45521/user/jobmanager.
2016-06-13 07:28:22,829 INFO  org.apache.flink.yarn.ApplicationClient                       - Successfully registered at the ResourceManager using JobManager Actor[akka.tcp://flink@10.0.2.15:45521/user/jobmanager#1868998090]
TaskManager status (0/1)
TaskManager status (0/1)
TaskManager status (0/1)
TaskManager status (0/1)
TaskManager status (0/1)
TaskManager status (0/1)
TaskManager status (0/1)
TaskManager status (0/1)
All TaskManagers are connected
Cluster started
Using address /10.0.2.15:45521 to connect to JobManager.
JobManager web interface address http://quickstart.cloudera:8088/proxy/application_1447844011707_0039/
Starting execution of program
Executing WordCount example with default input data set.
Use --input to specify file input.
Printing result to stdout. Use --output to specify output path.

------------------------------------------------------------
 The program finished with the following exception:

Job was submitted in detached mode. Results of job execution, such as accumulators, runtime, job id etc. are not available. Please make sure your program doesn't call an eager execution function [collect, print, printToErr, count]. 
	org.apache.flink.client.program.DetachedEnvironment$DetachedJobExecutionResult.getAccumulatorResult(DetachedEnvironment.java:103)
	org.apache.flink.api.java.DataSet.collect(DataSet.java:412)
	org.apache.flink.api.java.DataSet.print(DataSet.java:1605)
	org.apache.flink.examples.java.wordcount.WordCount.main(WordCount.java:92)
2016-06-13 07:28:27,206 INFO  org.apache.flink.yarn.YarnClusterClient                       - Disconnecting YarnClusterClient from ApplicationMaster
2016-06-13 07:28:27,208 INFO  org.apache.flink.yarn.ApplicationClient                       - Stopped Application client.
2016-06-13 07:28:27,208 INFO  org.apache.flink.yarn.ApplicationClient                       - Disconnect from JobManager Actor[akka.tcp://flink@10.0.2.15:45521/user/jobmanager#1868998090].
2016-06-13 07:28:27,307 INFO  org.apache.flink.yarn.YarnClusterClient                       - Application application_1447844011707_0039 finished with state RUNNING and final state UNDEFINED at 0
[cloudera@quickstart build-target]
```

In the JobManager logs, there is no sign of a job being submitted.

rmetzger commented on docs/setup/yarn_setup.md:
```diff
@@ -143,6 +143,34 @@ Note that in this case its not possible to stop the YARN session using Flink.
+~~~bash
+Usage:
+   Required
+     -id,--applicationId <yarnAppId> YARN application Id
```
I think this option is also not listed when running ./bin/yarn-session.sh.

mxm replied:
Yes, same as above. Fixed.

rmetzger commented:
Thanks a lot for adding this feature. I looked through the code and tried it out on a cluster.
Once my concerns are addressed (either through changes or comments) we can merge the change.

mxm commented:
Thanks for the review! I've addressed almost all your comments and will ping you once the PR get updated. Some more comments: 

> (1) When listing the yarn-session.sh cli options, I could not see the newly added argument

Fixed. The argument was not printed. The code now prints all available options.

>Starting a YARN session works. Also, connecting with a second YARN session works. However, if I'm stopping the session from the second session client, the first one does not notice this.
It would be good if all connected clients receive the messages, instead of the last one connected.

Yes, would be nice to have that feature but I think it is out of the scope of this PR.

>Detached per job submission is not working (./bin/flink run -m yarn-cluster -yd -yn 1 ./examples/batch/WordCount.jar)

You're running a WordCount with interactive job submission detached. That doesn't work because the plan can't be extracted in interactive programs (programs which use count/collect/print).

>While trying out your code, I noticed that the .yarn-properties file is not properly deleted, even though my previous yarn session was shutting down correctly:

I'll look into it. I know that this is also an issue in the current master. I had the properties file lingering around many times. I think it is not properly cleaned up in the ShutdownHook.

mxm commented:
I've addressed your comments in another commit. I'd be delighted if you could try out the changes again! 

- remove legacy code from CliFrontend
- change activation code of CustomCommandLine interface
- use checked exceptions to signal supported operations
- remove all checked exceptions of type Exception
- fix proper handling of yarn properties file
- fix logging and reduce verbosity of per-job clusters
- print 'id' argument in YarnSessionCli
- minor renaming of methods names
- improve documentation
- deprecate streaming option
- deprecate -m 'yarn-cluster'

rmetzger commented:
Thanks a lot for addressing my comments to quickly.
I'll try the code out soon.

mxm commented:
Rebased to the latest master and refined the tests.

rmetzger commented:
I tried your code again on my testing machine and everything that didn't work at my last tests is now fixed.

I also quickly scanned the newly added commits.
I think the pull request is good to merge.

mxm commented:
Merged. Thank you for your comments!
